keps/993-two-phase-admission/README.md

```diff
@@ -0,0 +1,226 @@
+to ensure that the resources can be provided in full before starting the workload.  
+
+#### Story 2
+
+I want to base workload admission on budget. I have the actual consumption metrics
+in my Prometheus/Datadog/Stackdriver/whatever monitoring system. And weekly
+budget in a CRD, per namespace. I’m happy to modify some small existing sample
+controller, but I don’t want to fork the whole Kueue to adjust some minor details.
+ 
+### Notes/Constraints/Caveats (Optional)
+
+Each of the go/no-go decisions is best-effort and can be reversed at any time. 
+
+### Risks and Mitigations
+
+## Design Details
+
+We will extend ClusterQueueSpec definition by adding a reference to
+AdmissionChecks that the queue needs to perform before admitting a
+workload. The design of AdmissionChecks is similar to IngressClass
+- it allows passing additional parameters, specific to particular checks
+that need to be performed. 
+
+```
+// ClusterQueueSpec defines the desired state of ClusterQueue
+type ClusterQueueSpec struct {
+[...]
+	// List of AdmissionCheck names that needs to be passed
+	// to admit a workload.  
+	AdmissionChecks []string `json:"addotomalChecks"`
+}
+
+
+// Cluster-scoped top-level object defining a check that 
+// can be referenced in ClusterQueue
+type AdmissionCheck struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec AdmissionCheckSpec `json:"spec,omitempty"`
+}
+
+const (
+	Retry CheckFailureState = “Retry”
+	Reject CheckFailureState = “Reject”
+)
+
+type AdmissionCheckSpec struct {
+// Name of the controller which will actually perform
+// the checks. Cannot be empty.
+	ControllerName string `json:"controllerName"`
+
+	// How long to keep the workload suspended 
+	// after a failed check.
+	RetryDelayMinutes *int64. `json:"retryDelayMinutes,omitempty"`
+
+	// A reference to the additional parameters for the check. 
+	Parameters *AdmissionCheckParametersReference `json:"parameters,omitempty"`
+}
+
+// Points to where the parameters for the checks are. 
+// As clusterqueue are in cluster scope - this should be
+// a dedicated CRD specific for the Controller.
+type AdmissionCheckParametersReference struct {
+	// ApiGroup is the group for the resource being referenced. 
+	APIGroup string `json:"apiGroup"`
+
+	// Kind is the type of the resource being referenced.
+	Kind string `json:"kind""`
+
+	// Name is the name of the resource being referenced.
+	Name string `json:"name"`
+}
+```
+
+For every workload that is put to a ClusterQueue, Kueue will perform the very 
+same checks that it does today, before admitting a workload. However, once the basic checks pass AND there are some AdmissionChecks configured, AND the workload is not in on-hold retry state from some check, it will:
+
+1. Fill the Admission field in workload, with the desired flavor assignment. 
+2. Not do any preemptions yet.
```

Comment:
This one is tricky. Let's take the scenario of ProvisioningRequest.

If a Workload requires preemptions to fit and we don't do preemptions, the ProvisioningRequest would likely trigger a scale up or fail because it reached the scale up limits (or reached stockout).

OTOH, if we do the preemptions, and the ProvisioningRequest cannot be satisfied for whatever reason, we would have done preemptions for nothing (or too early). Not sure if this should be a configuration parameter for the check.

Another open question: if we defer preemptions, should any more workloads be admitted in the cohort while waiting for the admissionChecks? In general I think we should not. Or perhaps it should be per flavor/resource that needs preemption or is borrowing, like in #1036.

Reply:
+1 for this being tricky, and also not very easy to do with the current preemption mechanism.
On the overall two steps admission, I propose, as a first stage to not change the way the preemption is done.

In the future, maybe we can add a built in `AdmissionCheckPreempt`, the controller of which should be in charge of issuing evictions and wait for the resources to become available.

Reply:
We cannot start with preemptions - if we preempt we will create a "hole" in in the quota and the preempted workload will try to get in again immediately. Thus preemption can only work with simultaneous admission or other form of the resource reservation. Thus if we preeempt, we should account the precheckedWorkloads in the quota, that basically means that we cannot have best effort queues as we have right now, only "moderate effort" ;).

Reply:
In general, we probably need 3 modes of operations:

* Preempt after checks pass - for checks that are not fully correlated with quota, and can take a longer time to give a green light.
* Preempt before checks are started - for checks that are essentially yet another quota check.
* Mix of the two based on the status provided by the check. The check may say:
   * Preempt now, as i'm hitting quota issues and preemption may help
   * Preempt after I give a green light, no quota issues observed at this moment and the admission will take a longer time, so there is no need to stop the running workloads yet. 

I will update the document to reflect that.